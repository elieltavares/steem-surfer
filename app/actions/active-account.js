/*
eslint-disable camelcase
 */

import { setItem, removeItem } from '../helpers/storage';
import { getAccounts, getAccountRC } from '../backend/steem-client';

import { getPoints } from '../backend/esteem-client';

export const LOGGED_IN = 'active-account/LOGGED_IN';
export const LOGGED_OUT = 'active-account/LOGGED_OUT';
export const UPDATED = 'active-account/UPDATED';

export const logIn = username => (dispatch, getState) => {
  setItem(`active_account`, username);
  dispatch(loggedIn(username));
  update(username, dispatch, getState);
};

export const logOut = () => dispatch => {
  removeItem(`active_account`);
  dispatch(loggedOut());
};

export const updateActiveAccount = () => (dispatch, getState) => {
  const { activeAccount } = getState();
  const { username } = activeAccount;
  update(username, dispatch, getState);
};

const update = (username, dispatch, getState) => {
  getAccounts([username])
    .then(resp => {
      const account = resp[0];

      return getAccountRC(username).then(r => {
        if (r.rc_accounts && r.rc_accounts.length > 0) {
          return Object.assign({}, account, { rcAccount: r.rc_accounts[0] });
        }

        return account;
      });
    })
    // Synchronising user's blog for reblog detection causes big amount of data usage. disabled.
    // See components/elements/EntryReblogBtn.js:43
    .then(account => Object.assign({}, account, { blog: [] }))
    .catch(() => {})
    .then(account =>
      getPoints(username).then(p => {
        const o = Object.assign({}, account, {
          unclaimed_points: p.unclaimed_points
        });

        // check active user still same
        const { activeAccount: a } = getState();
        const { username: u } = a;

        if (a && u === username) {
          dispatch(updated(username, o));
        }
        return o;
      })
    )
    .catch(() => {});
};

export const loggedIn = username => ({
  type: LOGGED_IN,
  payload: { username }
});

export const loggedOut = () => ({
  type: LOGGED_OUT,
  payload: {}
});

export const updated = (username, accountData) => ({
  type: UPDATED,
  payload: { username, accountData }
});
