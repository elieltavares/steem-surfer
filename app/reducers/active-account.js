import { getItem } from '../helpers/storage';

import { LOGGED_IN, LOGGED_OUT, UPDATED } from '../actions/active-account';

import { ACCOUNT_UPDATED_SC } from '../actions/accounts';

const load = () => {
  const username = getItem('active_account');
  if (username && getItem(`account_${username}`)) {
    return Object.assign({}, getItem(`account_${username}`), {
      accountData: null
    });
  }

  return null;
};

const defaultState = load();

export default function activeAccount(state = defaultState, action) {
  switch (action.type) {
    case LOGGED_IN:
    case LOGGED_OUT: {
      return load();
    }
    case UPDATED: {
      const { accountData } = action.payload;
      return Object.assign({}, state, { accountData });
    }
    case ACCOUNT_UPDATED_SC: {
      const { username, accountData } = action.payload;
      // active user updated
      if (state.username === username) {
        const { accessToken, refreshToken } = accountData;
        return Object.assign({}, state, { accessToken, refreshToken });
      }
      return state;
    }
    default:
      return state;
  }
}
