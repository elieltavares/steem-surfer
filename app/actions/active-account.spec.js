/* eslint-disable */
import * as actions from './active-account';

describe('active account actions', () => {
  it('(1) loggedIn action creator', () => {
    expect(actions.loggedIn('foo')).toMatchSnapshot();
  });

  it('(2) loggedOut action creator', () => {
    expect(actions.loggedOut()).toMatchSnapshot();
  });

  it('(3) updated action creator', () => {
    expect(actions.updated('foo', { id: 12, name: 'foo' })).toMatchSnapshot();
  });
});
