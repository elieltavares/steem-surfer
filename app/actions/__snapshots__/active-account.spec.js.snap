// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`active account actions (1) loggedIn action creator 1`] = `
Object {
  "payload": Object {
    "username": "foo",
  },
  "type": "active-account/LOGGED_IN",
}
`;

exports[`active account actions (2) loggedOut action creator 1`] = `
Object {
  "payload": Object {},
  "type": "active-account/LOGGED_OUT",
}
`;

exports[`active account actions (3) updated action creator 1`] = `
Object {
  "payload": Object {
    "accountData": Object {
      "id": 12,
      "name": "foo",
    },
    "username": "foo",
  },
  "type": "active-account/UPDATED",
}
`;
